File: push/task_list.go
package push

import (
	"fmt"
	"io/fs"
	"path/filepath"
	"strings"

	"github.com/aura-studio/dynamic-node-cli/config"
)

// TaskList holds push tasks grouped by remote.
type TaskList struct {
	Tasks map[string][]Pair
}

// Pair represents a local-to-remote file mapping.
type Pair struct {
	RemoteFilePath string
	LocalFilePath  string
}

// Add appends a task to the list for a given remote.
func (f *TaskList) Add(remote string, remoteFilePath string, localFilePath string) {
	f.Tasks[remote] = append(f.Tasks[remote], Pair{
		RemoteFilePath: remoteFilePath,
		LocalFilePath:  localFilePath,
	})
}

// NewTaskList creates a TaskList by scanning the warehouse for .zip artifacts.
func NewTaskList(proc config.Procedure) *TaskList {
	fileList := &TaskList{Tasks: make(map[string][]Pair)}

	name := proc.Target.Namespace + "_" + proc.Target.Package + "_" + proc.Target.Version
	env := proc.Toolchain.OS + "_" + proc.Toolchain.Arch + "_" + proc.Toolchain.Compiler + "_" + proc.Toolchain.Variant
	dir := filepath.Join(proc.Warehouse.Local, env, name)

	// Expected filename: libnode_<name>.zip
	libnodeName := fmt.Sprintf("libnode_%s.zip", name)

	for _, remote := range proc.Warehouse.Remote {
		if err := filepath.WalkDir(proc.Warehouse.Local, func(path string, d fs.DirEntry, err error) error {
			if err != nil {
				return err
			}
			if d.IsDir() {
				return nil
			}
			// Only consider files inside the expected dir
			if !strings.HasPrefix(path, dir) {
				return nil
			}
			base := filepath.Base(path)
			// Match primary artifact and timestamped backups
			if base == libnodeName || strings.HasPrefix(base, libnodeName+".") {
				rel, relErr := filepath.Rel(proc.Warehouse.Local, path)
				if relErr != nil {
					return relErr
				}
				remotePath := filepath.ToSlash(rel)
				fileList.Add(remote, remotePath, path)
			}
			return nil
		}); err != nil {
			panic(err)
		}
	}

	return fileList
}

File: src/push/task-list.js
import fs from "node:fs";
import path from "node:path";

export class TaskList {
  constructor() {
    this.tasks = new Map();
  }

  add(remote, remoteFilePath, localFilePath) {
    const tasks = this.tasks.get(remote) ?? [];
    tasks.push({ remoteFilePath, localFilePath });
    this.tasks.set(remote, tasks);
  }
}

export async function newTaskList(proc) {
  const fileList = new TaskList();

  const name = `${proc.target.namespace}_${proc.target.package}_${proc.target.version}`;
  const environment = [
    proc.toolchain.os,
    proc.toolchain.arch,
    proc.toolchain.compiler,
    proc.toolchain.variant,
  ].join("_");
  const dir = path.join(proc.warehouse.local, environment, name);
  const libnodeName = `libnode_${name}.zip`;

  for (const remote of proc.warehouse.remote) {
    await walkFiles(proc.warehouse.local, async (filePath) => {
      if (!filePath.startsWith(dir)) {
        return;
      }

      const base = path.basename(filePath);
      if (base === libnodeName || base.startsWith(`${libnodeName}.`)) {
        const rel = path.relative(proc.warehouse.local, filePath);
        const remotePath = toRemotePath(rel);
        fileList.add(remote, remotePath, filePath);
      }
    });
  }

  return fileList;
}

async function walkFiles(root, visit) {
  const entries = await fs.promises.readdir(root, { withFileTypes: true });

  for (const entry of entries) {
    const current = path.join(root, entry.name);
    if (entry.isDirectory()) {
      await walkFiles(current, visit);
      continue;
    }
    await visit(current);
  }
}

function toRemotePath(value) {
  return value.split(path.sep).join("/");
}
